### PR TITLE
Make torch.cuda.rng_set_state() and torch.cuda.rng_get_state() work during stream capture.

### LINKED ISSUE #162504: torch.utils.checkpoint.checkpoint does not work under stream capture to a cuda graph
### 🐛 Describe the bug

The following does not work today:

```python
import torch

torch.cuda.manual_seed(42)

def fn(x):
    return x * torch.sigmoid(torch.randn(1, device="cuda"))

# initialize device state
fn(torch.ones(1, device="cuda"))

torch.cuda.manual_seed(42)
eager_in = torch.ones(1, device="cuda", requires_grad=True)
eager_out = torch.utils.checkpoint.checkpoint(
    fn, eager_in,
    use_reentrant=False,
    preserve_rng_state=True,
)
eager_in_grad,  = torch.autograd.grad(eager_out, eager_in)

g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
    graph_in = torch.ones(1, device="cuda", requires_grad=True)
    graph_out = torch.utils.checkpoint.checkpoint(
        fn, graph_in,
        use_reentrant=False,
        preserve_rng_state=True,
    )
    graph_in_grad,  = torch.autograd.grad(graph_out, graph_in)

torch.cuda.manual_seed(42)
g.replay()
assert torch.allclose(eager_in_grad, graph_in_grad, rtol=0.0, atol=0.0), "Mismatch in gradient outputs"
print(eager_in_grad)
print(graph_in_grad)
```

It will fail with an assertion error from current_seed: https://github.com/pytorch/pytorch/blob/f03d635dc60068d63eda0d13e717036db60219cc/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp#L304

However, that is just the proximal cause of the issue. The real problem is that torch.cuda.get_rng_state() and torch.cuda.set_rng_state() are banned during stream capture, due to a variety of other assertions.

Even if you were to comment out all of the failing assertions in CUDAGeneratorImpl.cpp, the assertion in my above test case would fail because the rng state isn't being tracked correctly.

This matters because people who want to run training at speed of light in models that depend upon activation recomputation cannot do it right now if the training is cpu-bound.

I already have a fix. I will make a PR shortly. Just making this issue for documentation purposes.

### Versions

This problem exists on top of tree.

cc @soulitzer @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

### PR DESCRIPTION
**Pull request by galv**
Note that this works only in a limited case, where you *don't* change the seed, but change only the offset of the philox generator. This captures the main use case we're interested in: Rewinding the RNG to a previous state. This is done by torch.utils.checkpoint.checkpoint in particular.

Calls to increase() change only the offset, not the seed. Thus, we allow for "no-op" calls to set_seed where the new seed is the same as the old seed. If a user does happen to try to change the seed during stream capture, they will receive an error.

Fixes #162504

cc @pbelevich @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng

### CODE REVIEW

**Comment by ngimel** [comment]
cc @drisspg, this also would have been useful for sdpa with dropout, but oh well

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 1]
```diff
@@ -266,11 +266,14 @@ CUDAGeneratorImpl::CUDAGeneratorImpl(
  * See Note [Acquire lock when using random generators]
  */
 void CUDAGeneratorImpl::set_current_seed(uint64_t seed) {
-  at::cuda::assertNotCapturing(
-      "Cannot call CUDAGeneratorImpl::set_current_seed");
-  state_->seed_ = seed;
-  state_->philox_offset_per_thread_ = 0;
-  no_reset_rnn_state_.clear();
+  if (at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None) {
```
Comment (ngimel) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: C10_LIKELY

**Review by ngimel — APPROVED** [review]

**Review by eqy — APPROVED** [review]

**Review by eellison — APPROVED** [review]
Nice ! Would this work for the use case in https://github.com/pytorch/pytorch/pull/114068 as well ? cc @eee4017

**Review by eee4017 — APPROVED** [review]

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 2]
```diff
@@ -346,8 +346,6 @@ c10::intrusive_ptr<c10::TensorImpl> CUDAGeneratorImpl::get_state() const {
  * and size of the internal state.
  */
 void CUDAGeneratorImpl::set_state(const c10::TensorImpl& new_state) {
-  at::cuda::assertNotCapturing(
```
Comment (eee4017) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:349]: I think we only want to use this PR when we don’t want the offset to increase during the replay stage, right? #114068 was mainly to ensure that offset exchange remain safe during replay (See discussion here #113541). If the user isn’t aware that the offset is not updated in replay and calls `CUDAGeneratorImpl::set_state` during capture, they could end up with incorrect results.

Comment (galv) [reply aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:349]: Yeah, the seeds could also be different among the states that get exchanged.

I initially considered using your API, BTW, but it was intrusive to integrate because I needed to detect whether stream capture was happening any time the user might want to call torch.cuda.get_rng_state. If stream capture was happening, I would call the graph safe getters and setters. So I could not make that work for arbitrary user code.

Comment (eee4017) [reply aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:349]: 1. For the code here, I means that should you have any warning/reminder for the user if they use the set_state in the capture and get incorrect results unexpectedly, it should be hard for them to debug if there is no assertion.
2. Why do you need to check the capture status when using the graph safe set state? I think we can use it without the capturing.

**Review by cyyever — APPROVED** [review]
